Calibration/EcalCalibAlgos/plugins/EcalPhiSymRecHitProducers.cc

```diff
@@ -0,0 +1,574 @@
+struct PhiSymCache {
+    mutable tbb::concurrent_vector<EcalPhiSymInfo> ecalLumiInfo;
+    mutable tbb::concurrent_vector<EcalPhiSymRecHit> recHitCollEB;
+    mutable tbb::concurrent_vector<EcalPhiSymRecHit> recHitCollEE;
+
+  void clear() {
+      ecalLumiInfo.clear();
+    recHitCollEB.clear();
+    recHitCollEE.clear();
+  }
+};
+
+//---Wrapper to handle stream data
+struct PhiSymStreamCache {
+    mutable EcalPhiSymInfo ecalLumiInfo;
+    mutable EcalPhiSymRecHitCollection recHitCollEB;
+    mutable EcalPhiSymRecHitCollection recHitCollEE;
+
+  void clear() {
+      ecalLumiInfo = EcalPhiSymInfo();
+    recHitCollEB.clear();
+    recHitCollEE.clear();
+  }
+};
+
+//****************************************************************************************
+// - EcalPhiSymRecHitProducerBase: base class implementing the main algorithm
+// - EcalPhiSymRecHitProducerLumi: produces reduced collections per LuminosityBlock
+// - EcalPhiSymRecHitProducerRun: produces reduced collections per Run
+class EcalPhiSymRecHitProducerBase {
+public:
+  explicit EcalPhiSymRecHitProducerBase(const edm::ParameterSet& pSet, edm::ConsumesCollector&& cc);
+  ~EcalPhiSymRecHitProducerBase(){};
+
+  //---methods
+  // job
+  void initializeJob();
+  // event
+  void processEvent(edm::Event const& event, edm::EventSetup const& setup, PhiSymStreamCache* cache) const;
+  // helpers
+  void initializeStreamCache(PhiSymStreamCache* cache) const;
+  void initializeGlobalCache(edm::EventSetup const& setup,
+                             edm::ESGetToken<EcalChannelStatus, EcalChannelStatusRcd> const& chStatusToken,
+                             edm::ESGetToken<CaloGeometry, CaloGeometryRecord> const& geoToken,
+                             std::shared_ptr<PhiSymCache>& cache) const;
+  void sumCache(PhiSymStreamCache* streamCache, PhiSymCache const* cache) const;
+
+  //---data memebers
+  // available to derived classes
+protected:
+  mutable std::vector<DetId> barrelDetIds_;
+  mutable std::vector<DetId> endcapDetIds_;
+  edm::ESGetToken<CaloGeometry, CaloGeometryRecord> geoToken_;
+  edm::ESGetToken<EcalLaserDbService, EcalLaserDbRecord> laserDbToken_;
+  mutable edm::EDGetTokenT<EBRecHitCollection> ebToken_;
+  mutable edm::EDGetTokenT<EBRecHitCollection> eeToken_;
+  float etCutEB_;
+    std::vector<double> eThresholdsEB_;
+  float etCutEE_;
+  std::vector<double> A_;
+  std::vector<double> B_;
+  float thrEEmod_;
+  int nMisCalib_;
+  int nSumEtValues_;
+  std::vector<double> misCalibRangeEB_;
+  std::vector<float> misCalibStepsEB_;
+  std::vector<double> misCalibRangeEE_;
+  std::vector<float> misCalibStepsEE_;
+  //---geometry
+  EcalRingCalibrationTools calibRing_;
+  static const short kNRingsEB = EcalRingCalibrationTools::N_RING_BARREL;
+  static const short kNRingsEE = EcalRingCalibrationTools::N_RING_ENDCAP;
+  static const short ringsInOneEE = kNRingsEE / 2;
+  mutable float etCutsEB_[kNRingsEB];
+  mutable float etCutsEE_[kNRingsEE];
```

Comment:
I understand why these are not in principle thread safe, but actually they are never re-written after the first call to `initializeGlobalCache` the arrays are initialized to -1 in `beginJob` and then the correct value is set in the first call of `initializeGlobalCache` and never touched again. The only reason I have to do this is that the geometry is not available during `beginJob`, we do not expect the ECAL geometry to change across a run. Do you think it is anyway wiser to put these into the Lumi/Run cache?

Reply:
Maybe I should just make all of these arrays/vectors, array/vector of std::atomic, to ensure that different streams are not trying to write at the same time and ensure the writing happens only once

Reply:
As far as I can tell the `initializeGlobalCache()` (now `initializePhiSymCache()`) is called on each `globalBeginLuminosityBlock()` / `globalBeginRun()`, and I didn't see any check if the values have been filled already. Did I miss something?

The mental model is that `globalBeginLuminosityBlock()` and `globalBeginRun()` can be called concurrently for different Lumis/Runs (it may be that today the system incidentally serializes some things such that some test can succeed, but following the model will cause less problems in the future).

The most correct approach would be to put these into EventSetup, then the IOVs would be guaranteed to be handled properly. If that feels overkill, these can be set within the module, and that must be done in a thread safe way. The "only one geometry in a job" is an assumption done in many places, but with my `core` hat I can't recommend it. Within a module the best way to handle data derived from EventSetup is to put it in `LuminosityBlockCache` (and using `edm::ESWatcher` + `edm::ReusableObjectHolder` + `edm::SerialTaskQueue` to update them only when the IOV actually changes). A `RunCache` could work in practice too if you're sure that the IOV will never be shorter than Run.

I don't recommend to use vector/array of `std::atomic` if there is another way and algorithm isn't really about atomic updates, because the code becomes more difficult to reason about.